src/ApiCore/ResourceTemplate/RelativeResourceTemplate.php

```
@@ -0,0 +1,296 @@
+    {
+        return new ValidationException("Could not match path '$path' to template '$this'");
+    }
+
+    /**
+     * @param Segment[] $segments
+     * @return array[] A list of [string, Segment] tuples
+     */
+    private static function buildKeySegmentTuples($segments)
+    {
+        $keySegmentTuples = [];
+        $positionalArgumentCounter = 0;
+        foreach ($segments as $segment) {
+            switch ($segment->getSegmentType()) {
+                case Segment::WILDCARD_SEGMENT:
+                case Segment::DOUBLE_WILDCARD_SEGMENT:
+                    $positionalKey = "\$$positionalArgumentCounter";
+                    $positionalArgumentCounter++;
+                    $keySegmentTuples[] = [$positionalKey, $segment];
+                    break;
+                default:
+                    $keySegmentTuples[] = [$segment->getKey(), $segment];
+            }
+        }
+        return $keySegmentTuples;
+    }
+
+    /**
+     * @param array[] A list of [string, Segment] tuples
+     * @return array[] A list of [string, Segment] tuples
+     */
+    private static function flattenKeySegmentTuples($keySegmentTuples)
+    {
+        $flattenedKeySegmentTuples = [];
+        foreach ($keySegmentTuples as list($key, $segment)) {
+            /** @var Segment $segment */
+            switch ($segment->getSegmentType()) {
+                case Segment::VARIABLE_SEGMENT:
+                    // For segment variables, replace the segment with the segments of its children
+                    $template = $segment->getTemplate();
+                    $nestedKeySegmentTuples = self::buildKeySegmentTuples($template->segments);
+                    foreach ($nestedKeySegmentTuples as list($nestedKey, $nestedSegment)) {
+                        /** @var Segment $nestedSegment */
+                        // Nested variables are not allowed
+                        assert($nestedSegment->getSegmentType() !== Segment::VARIABLE_SEGMENT);
```

It's possible for users to disable/modify the behavior of `assert` calls. If the intent here is to prevent further execution, we may need to throw an exception. (This applies to other usage of assert as well)

---

This is deliberate. All of the asserts are things that should never happen, regardless of user input - if they happen, it is a bug in the code. So they are there so be checked during testing, but users do not need them to validate their inputs.

---

Sounds good, I like that